### ISSUE TITLE
Frequent reloads result in high CPU usage

### ISSUE THREAD

**Issue by mateuszdrab** [haproxytech/kubernetes-ingress#514]
Hi guys

Lately, I've been observing a pattern of frequent Haproxy reloads, increasing with frequency until the CPU usage grows so high, a node 'dies'. 
Below are two charts, one showing aggregated CPU usage for Haproxy and one showing the count of 'reload required' words in the logs.
![image](https://user-images.githubusercontent.com/26556541/211332367-5efc8df8-6195-4ee9-bdad-2450af5439dc.png)
![image](https://user-images.githubusercontent.com/26556541/211332145-f2f53009-5e08-4182-b335-4ea94123436e.png)

All of those reloads come with the following line:

```
DEBUG   rules/main.go:168 New HAProxy rule 'REQ_DENY' created, reload required
```
I'm not sure where this 'rule' is coming from - I'm not adding anything to the cluster that involves ingresses or haproxy.

Does this sound like it could be some loop/race condition in the controller?

I also see a lot of events with the following line:

```
DEBUG   global.go:143 Defaults config updated: [ErrorFiles: <nil slice> != [0xc0045489f0]]
```
By a lot, I mean 1000+ over the past 24h.

I do have custom error files, and they work fine, but I have a feeling that the situation here might be similar to https://github.com/haproxytech/kubernetes-ingress/issues/459

The `<nil slice>` does not sound right, could a comparison be in place that thinks the error files changed when they had not, resulting in constant reloads?

Thanks

**Comment by mateuszdrab** [comment]
After disabling the error files argument, the situation improved significantly, a small count of REQ_DENY messages but no more 'nil slice' messages.

![image](https://user-images.githubusercontent.com/26556541/211350675-87147a1f-210f-4c6e-a52f-d2de8e381472.png)

**Comment by mateuszdrab** [comment]
I'm continuing to face the issue, I've been thinking perhaps it's due to load, but then it's also happening on my 2 least used nodes, which experience little amount of connections (at most 10/s). It looks like the CPU use grows from start of the pod until it saturates the node (there are no limits).

**Comment by ivanmatmati** [comment]
Hi @mateuszdrab , Thanks for reporting, we'll look at it.

**Comment by ivanmatmati** [comment]
Can you also provide the logs for the peaks periods ?

**Comment by mateuszdrab** [comment]
Sure, I have experienced a few more peaks which are not correlatable with actual load as load remains stable and CPU usage and observed reloads increase.
I'm quite interested in where this REQ_DENY rule is coming from.

I'll get some logs and post them shortly

**Comment by mateuszdrab** [comment]
![image](https://user-images.githubusercontent.com/26556541/211633709-f2582635-e3ce-485d-84c4-968d0adff8fe.png)
Not sure if you can tell, the wide bars are the address not available messages and the small bars are the REQ_DENY rule messages.

**Comment by ivanmatmati** [comment]
Can you also check you have white or black list present in your configuration ?

**Comment by mateuszdrab** [comment]
> Thanks, you didn't provide the version of the controller you're using. Can you ?

Latest, 1.9.0 thanks to my friend FluxCD 🤓

> Can you also check you have white or black list present in your configuration ?

Yes, two ingresses have the whitelist annotation (out of a total of ~30)

- The situation has been very calm today, no crashes; however, I still have the error pages configmap disabled...
- I also resolved an issue today where an ingress (based by externalname) service was pointing to a dns name which pointed back at the ingress. I did it today at 1pm and observing the reloads, there seems to be no big difference. 
- Since yesterday, I excluded 2 haproxy instances (which are on nodes with a slow link) from the main service. I am using `Internal/ExternalTrafficPolicy` as `Local`, but it seems that this does not affect cluster generated traffic sent to the LB IP. This resulted in lots of traffic going to those slow instances, and those were crashing a lot.

Some questions:
- Do haproxy instances talk to each other? Could instability of one instance affect others (for example by causing reloads)?
- Is there a reload metric? I don't see one so I generate my stats by parsing logs, but a prom metric would be very useful here, especially if it could carry a label indicating the source/trigger for the reload.

**Comment by ivanmatmati** [comment]
Haproxy instances don't communicate with each other. But there could be some side effect from one to another. For example, when using public service the status of managed ingressed is modified by an update sent to API server. The side effect is that an update of status will trigger a modification event sent to all instances.
Concerning metric , I don't think there's a metric about reload but I'll ask the team.
On my side I've seen an issue in the manner we deal with error files state in the comparison made for the reload decision.
For the REQ_DENY case, I didn't see anything special in my test with black and white lists. Can you tell me if you're using pattern files or direct insertion for these lists? If there's a reload it's because there's an event triggering it. I didn't see reload happening for nothing in my tests. If nothing happens no reload occurs seems to be still valid. Can you check in your k8s events what could have triggered the reloads ?

**Comment by mateuszdrab** [comment]
@ivanmatmati 

Thanks for the response, my environment has been stable since I applied the fixes mentioned in my previous post.
I've since been able to re-deploy the error pages and the reloads went up slightly, but this is pretty much not noticeable.
I do think something is wrong with the reloads when the error files are in place, hopefully this will get resolved.
I used to face a lot of disconnection issues, on long running 'watch' style operations (but happening very shortly after the operation is started) - I made an assumption that the haproxy reloads are the cause of those but never confirmed. 
Do reloads cause existing connections to be closed?

>Can you tell me if you're using pattern files or direct insertion for these lists?
Nothing of that sort, the annotation is all I use.

This is a chart of reload count due to REQ_DENY over 5m
![image](https://user-images.githubusercontent.com/26556541/212665954-48eef341-330a-4ec3-9e8f-928a9beca2b5.png)
